scripts/docs_link_checker.py

```diff
@@ -0,0 +1,306 @@
+import glob
+import logging
+import os
+import re
+from typing import List
+
+import click
+import requests
+
+logger = logging.getLogger()
+logger.addHandler(logging.StreamHandler())
+logger.setLevel(logging.INFO)
+
+
+class LinkReport:
+    def __init__(self, link: str, file: str, message: str):
+        self.link = link
+        self.file = file
+        self.message = message
+
+    def __str__(self):
+        return f"{self.message}: File: {self.file}, Link: {self.link}"
+
+
+class LinkChecker:
+    def __init__(
+        self, docs_path: str, docs_root: str, site_prefix: str, skip_external: False
+    ):
+        self._docs_path = docs_path.strip(os.path.sep)
+        self._docs_root = docs_root.strip(os.path.sep)
+        self._site_prefix = site_prefix.strip("/")
+        self._skip_external = skip_external
+
+        markdown_link_regex = r"!?\[(.*?)\]\((.*?)\)"  # inline links, like [Description](link), images start with !
+        self._markdown_link_pattern = re.compile(markdown_link_regex)
+
+        external_link_regex = r"^https?:\/\/"  # links that start with http or https
+        self._external_link_pattern = re.compile(external_link_regex)
+
+        # links that being with /{site_prefix}/(?P<path>), may end with #abc
+        absolute_link_regex = r"^\/" + site_prefix + r"\/(?P<path>[\w\/-]+?)(?:#\S+)?$"
+        self._absolute_link_pattern = re.compile(absolute_link_regex)
+
+        # docroot links start without a . or a slash
+        docroot_link_regex = r"^(?P<path>\w[\.\w\/-]+\.md)(?:#\S+)?$"
+        self._docroot_link_pattern = re.compile(docroot_link_regex)
+
+        # links starting a . or .., file ends with .md, may include an anchor with #abc
+        relative_link_regex = r"^(?P<path>\.\.?[\.\w\/-]+\.md)(?:#\S+)?$"
+        self._relative_link_pattern = re.compile(relative_link_regex)
+
+        absolute_image_regex = r"^\/" + site_prefix + r"\/(?P<path>[\w\/-]+\.\w{3,4})$"
+        self._absolute_image_pattern = re.compile(absolute_image_regex)
+
+        # ending with a 3-4 character suffix
+        relative_image_regex = r"^(?P<path>\.\.?[\.\w\/-]+\.\w{3,4})$"
+        self._relative_image_pattern = re.compile(relative_image_regex)
+
+    def _is_image_link(self, markdown_link: str):
+        return markdown_link.startswith("!")
+
+    def _is_doc_link(self, markdown_link: str):
+        return not self._is_image_link(markdown_link)
+
+    def _is_anchor_link(self, link: str):
+        return link.startswith("#")
```

Comment:
Could we add type hints for return values here? Perhaps the `_is` prefix is good enough to represent a `bool` but consistency is nice 😄

Reply:
done!